### PR TITLE
Simplify the handling of the parallel processing in exact mode

### PR DESCRIPTION
**Pull request by astrofrog**
*may* solve #76

### CODE REVIEW

**Comment by astrofrog** [comment]
@cdeil - looks like Travis is happy so I will plan to merge this anyway, but curious whether it solves your issue?

**Comment by cdeil** [comment]
@astrofrog – I just got back to this ... still testing.

I've managed to boil down the testcase to this:
https://gist.github.com/cdeil/709802defe5fc9f50177

Put this in an IPython notebook cell. Execute it once – OK. Execute it again - Crash.
Remove the `imshow` line – never crashes.

This only happens in the IPython notebook ... it must have to do with IPython and / or reproject making assumptions about the input data array that are invalid at the C level. My guess this is a bug in reproject.
I'll test this branch now.

You still can't reproduce with reproject master?

**Comment by cdeil** [comment]
I'm getting the same crash with this branch.
@astrofrog – Do you want me to continue debugging on this branch or master?

Any better idea than adding print statements to the reproject_exact code?

**Comment by astrofrog** [comment]
@cdeil - could you debug on this branch? The code is cleaner so I prefer it anyway. Thanks!

**Comment by astrofrog** [comment]
@cdeil - by the way, what version of IPython are you using? (just trying to also reproduce that here and failing).

**Comment by cdeil** [comment]
I'm using `py34-ipython @3.1.0_0` and `py34-matplotlib @1.4.3_0+cairo+pyside+webagg` from Macports on OS X 10.10 and Astropy and reproject master (or this branch).

**Comment by cdeil** [comment]
PS: I don't think this should stop the 0.1 release. Crashes are rare and scary for Python users, but probably it's unlikely that this will affect many users?

**Comment by astrofrog** [comment]
I can't reproduce the crash with the same versions but with conda. I will try MacPorts later. But it would be great if you could narrow down at what point in the code the crash happens. Can you check that this does not happen with ``parallel=False`` for ``reproject_exact``?

**Comment by cdeil** [comment]
I've never seen the crash with `parallel=False`.

I only get it from an IPython notebook with two calls to `reproject_exact` in different cells. If I move the second call in the first cell it doesn't crash.

I guess that's a strong hint that the issue has to do with parallel execution, but I'm not sure.

File: reproject/spherical_intersect/core.py [hunk 1]
```diff
@@ -130,69 +134,49 @@ def _reproject_celestial(array, wcs_in, wcs_out, shape_out, parallel=True, _meth
...
-            raise
-
-        finally:
-            if not pool is None:
-                # Clean up the pool, if still alive.
-                pool.close()
-                pool.join()
+        # Prime each process in the pool with a small function that disables
+        # the ctrl+c signal in the child process.
+        pool = Pool(nproc, _init_worker)
```
Comment (astrofrog) [inline reproject/spherical_intersect/core.py]: @cdeil - can you put print statements around here to see if this is where it's happening? This is where the processes are launched.

Comment (cdeil) [inline reproject/spherical_intersect/core.py]: I did put print statements.
The crash happens a few lines below: the `pool.close()` call in line 172.

**Comment by astrofrog** [comment]
@cdeil - by the time the processes are closed all the reproject code should be done, so this is very weird. Can you try running just:

```
def identity(x):
    return x
    
def do():
    
    import multiprocessing as mp
    
    pool = mp.Pool(processes=10)
    
    pool.map(identity, range(10))
    
    pool.close()
    
do()
do()
```

?

**Comment by cdeil** [comment]
Your simple example with `do()` never crashes.

**Comment by astrofrog** [comment]
@cdeil - just one quick favor, can you confirm that the following crashes for you?

http://nbviewer.ipython.org/gist/astrofrog/89e4c4b4889e8e412f56

**Comment by cdeil** [comment]
Yes, this does crash for me.
https://gist.github.com/cdeil/decb454a07ac871bf4d5

Do you understand this?
Where to file an issue about this?

**Comment by astrofrog** [comment]
See https://github.com/astrofrog/reproject/issues/76#issuecomment-99935672 - since happens only in the notebook, I'd say it's a notebook bug.

**Comment by astrofrog** [comment]
Also, I have NO IDEA what this is all about. But I'm happy that we have a MWE and it's not a reproject issue

**Comment by astrofrog** [comment]
In any case, I think this is good to go - it's an improvement over the existing code as it's simpler.

**Pull request merged**